hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/StorageBasedLockProvider.java

```diff
@@ -0,0 +1,581 @@
+   */
+  public StorageBasedLockProvider(final LockConfiguration lockConfiguration, final StorageConfiguration<?> conf) {
+    StorageBasedLockConfig config = new StorageBasedLockConfig.Builder()
+        .fromProperties(lockConfiguration.getConfig()).build();
+    heartbeatInterval = config.getHeartbeatPoll();
+    lockValidity = config.getLockValidityTimeout();
+
+    String configuredLocksLocation = config.getLocksLocation();
+
+    // If not configured, recalculate the locks location as .hoodie/.locks;
+    // otherwise (the lock location is configured), the configuration location is used as the folder
+    // to which the lock file is written to, and the lock file name is determined by the table's base path
+    String locksLocation = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? String.format("%s%s%s", config.getHudiTableBasePath(), StoragePath.SEPARATOR, LOCKS_FOLDER_NAME)
+        : configuredLocksLocation;
+
+    URI uri = parseURI(locksLocation);
+    bucketName = uri.getHost(); // For most schemes, the bucket/container is the host.
+    String folderName = uri.getPath(); // Path after the bucket/container.
+
+    String fileName = StringUtils.isNullOrEmpty(configuredLocksLocation)
+        ? DEFAULT_TABLE_LOCK_FILE_NAME
+        : slugifyLockFolderFromBasePath(config.getHudiTableBasePath());
+
+    lockFilePath = buildLockObjectPath(folderName, fileName);
+    ownerId = UUID.randomUUID().toString();
+    this.logger = LOGGER;
+    this.heartbeatManager = new LockProviderHeartbeatManager(
+        ownerId,
+        heartbeatInterval,
+        this::renewLock);
+
+    try {
+      this.lockService = (StorageLock) ReflectionUtils.loadClass(
+          getLockServiceClassName(uri.getScheme()),
+          new Class<?>[] { String.class, String.class, String.class, Properties.class },
+          new Object[] { ownerId, bucketName, lockFilePath, lockConfiguration.getConfig() });
+    } catch (Throwable e) {
+      throw new HoodieLockException("Failed to load and initialize StorageLock", e);
+    }
+    shutdownThread = new Thread(() -> shutdown(true));
+    Runtime.getRuntime().addShutdownHook(shutdownThread);
+    logger.info("Instantiated new storage-based lock provider, owner: {}, lockfilePath: {}", ownerId, lockFilePath);
+  }
+
+  private URI parseURI(String location) {
+    try {
+      return new URI(location);
+    } catch (URISyntaxException e) {
+      throw new HoodieLockException("Unable to parse locks location as a URI: " + location, e);
+    }
+  }
+
+  private static @NotNull String getLockServiceClassName(String scheme) {
+    Option<StorageSchemes> schemeOptional = StorageSchemes.getStorageLockImplementationIfExists(scheme);
+    if (schemeOptional.isPresent()) {
+      return schemeOptional.get().getStorageLockClass();
+    } else {
+      throw new HoodieNotSupportedException("No implementation of StorageLock supports this scheme: " + scheme);
+    }
+  }
+
+  @VisibleForTesting
+  StorageBasedLockProvider(
```

Comment:
shouldn't the other cnstr call this constructor?

Reply:
It's difficult to do this since we need to catch the errors